### PR TITLE
Make OpenCode custom models and agents selectable

### LINKED ISSUE #1093: Make OpenCode custom models and agents clear and selectable
## Problem

Users cannot tell how BB maps OpenCode models and agents into its execution controls.

BB already discovers ACP session models and applies a selected model with `session/set_model`. OpenCode custom models should therefore appear under `acp-opencode` when OpenCode advertises them.

However, several gaps make this flow unclear or incomplete:

- The BB guide does not explain how to select an OpenCode custom model.
- The guide does not explain the difference between an OpenCode model and an OpenCode agent.
- BB parses ACP config options, but it only uses `model` and `thought_level`.
- BB ignores the ACP `mode` option, which can represent OpenCode agents.
- The `customModels` fallback only accepts static provider IDs.
- That fallback rejects `acp-opencode` and other dynamic ACP provider IDs.

A user with a custom primary agent, such as an orchestrator, can mistake it for a model. BB gives no useful path for that case.

## Current code

- `packages/agent-runtime/src/acp/bridge/model-catalog.ts` maps ACP model options.
- `packages/agent-runtime/src/acp/bridge/bridge.ts` applies ACP model changes.
- `packages/config/src/bb-app-managed-config.ts` limits `customModels.providerId` to the static provider enum.
- `packages/templates/src/templates/bb-guide-providers.md` documents ACP providers, but not this selection flow.

## Proposed result

BB should give users one clear OpenCode selection flow.

- Show all custom models that OpenCode advertises.
- Let users select a custom OpenCode primary agent when ACP exposes it as a mode.
- Explain that models and agents are different selections.
- Add a supported fallback for a custom model that an ACP agent does not advertise.
- Make the CLI guide show the exact discovery commands.

Example:

```sh
bb provider models acp-opencode --environment "$BB_ENVIRONMENT_ID"
bb thread spawn --provider acp-opencode --model <provider/model>
```

## Acceptance criteria

- A custom OpenCode model appears in the BB model picker.
- The same model appears in `bb provider models acp-opencode`.
- A selected model reaches OpenCode before the first prompt.
- A custom OpenCode primary agent can be selected when ACP advertises it.
- The guide explains models, agents, discovery, and fallback behavior.
- Tests cover custom models, ACP mode selection, and dynamic ACP provider IDs.

## Related issues

Issue #365 covers model-list failure recovery. This issue covers discovery and selection.

Issue #1086 covers Pi extension models. This issue covers OpenCode and generic ACP controls.

### PR DESCRIPTION
**Pull request by SawyerHood**
OpenCode exposes models and primary agents as two separate ACP selections. BB previously hid the agent selection and dropped unadvertised custom models before session start.

This change makes both choices explicit. The model picker now has an Agent section when ACP advertises a mode option.

## What changed

- Discover ACP model and mode options from a temporary session.
- Send custom model selections through `session/set_model` before the first prompt.
- Send primary-agent selections through `session/set_config_option` before the first prompt.
- Add `bb provider modes` and `bb thread spawn --provider-mode`.
- Allow validated `acp-<slug>` provider IDs in `customModels`.
- Carry provider modes through the app, SDK contracts, server, host daemon, and ACP bridge.
- Increase the host daemon protocol version from 94 to 95.
- Document OpenCode discovery commands and fallback behavior.

## Protocol evidence

I checked OpenCode's current ACP implementation before I changed BB. OpenCode emits models in the `model` category and primary agents in the `mode` category.

OpenCode applies mode changes through `session/set_config_option`. It applies model changes through the ACP session model path.

The fake ACP fixture now matches those protocol shapes. The bridge tests prove that each selection takes effect before the first prompt.

## Validation

Before this change, `bb provider models acp-opencode --json` returned only the synthetic `acp-default` fallback in the reproduced environment.

After this change:

- The configuration tests accept dynamic ACP model IDs and reject invalid IDs.
- The CLI tests show the configured OpenCode model and the advertised primary agents.
- The ACP tests confirm custom model and mode selection before the first prompt.
- The full agent-runtime, CLI, host-daemon, server, app, contract, and template test suites pass.
- All affected Turbo type checks and builds pass.
- The app lint has no errors. Existing warnings remain.

Closes #1093

### CODE REVIEW

**Review by SawyerHood — COMMENTED** [review]
🚨 **SLOP COP** 🚨 · `review`

I am the SlopCop. I started this review under the security, code quality, performance, architecture, and end-to-end guidelines.

File: packages/agent-runtime/src/acp/bridge/bridge.ts [hunk 1]
```diff
@@ -1813,28 +1868,30 @@ async function handleRequest(
...
+          ? await loadSessionDiscoveredOptions(request.params.agent)
           : null;
-      if (sessionDiscoveredModels) {
+      if (sessionDiscoveredOptions) {
         sendResult(request.id, {
-          models: applyConfiguredReasoningToModels(sessionDiscoveredModels, {
-            reasoningCli: request.params.reasoningCli,
-            nativeReasoning: request.params.nativeReasoning,
-          }),
+          models: applyConfiguredReasoningToModels(
```
Comment (SawyerHood) [inline packages/agent-runtime/src/acp/bridge/bridge.ts]: 🚨 `slopcop/review` — Medium: keep the synthetic default model when an agent reports modes only.

An ACP agent can report modes without models. This branch then returns an empty model list.

The server rejects thread creation with `model_catalog_unavailable`. Return `ACP_DEFAULT_MODEL` when the discovered model list is empty.

Extend the mode discovery test to check both the model and the mode results.

File: packages/agent-runtime/src/acp/bridge/bridge.ts [hunk 2]
```diff
@@ -1070,6 +1077,42 @@ async function selectAcpNativeModel(args: {
...
+    !modeOption.options?.some((option) => option.value === args.providerMode)
+  ) {
+    throw new Error(
+      `The ACP agent does not offer provider mode "${args.providerMode}".`,
+    );
+  }
+  if (modeOption.currentValue === args.providerMode) {
+    return;
+  }
+  await args.connection.request({
```
Comment (SawyerHood) [inline packages/agent-runtime/src/acp/bridge/bridge.ts]: 🚨 `slopcop/review` — High: add a deadline to the new mode setup request.

If the ACP agent never answers, this bridge request waits after the outer runtime request fails at 120 seconds.

The bridge has not registered the session yet. Each retry can keep one child process and one request alive.

Cancel the request at the deadline. Close the connection and track incomplete sessions for shutdown.

File: packages/agent-runtime/src/acp/bridge/bridge.ts [hunk 3]
```diff
@@ -1813,28 +1868,30 @@ async function handleRequest(
...
+          request.params.primaryModels,
         );
+        sendResult(request.id, { ...models, modes: [] });
         return;
       }
-      const sessionDiscoveredModels =
+      const sessionDiscoveredOptions =
         request.params.listCommand === undefined && request.params.agent
-          ? await loadSessionDiscoveredModels(request.params.agent)
+          ? await loadSessionDiscoveredOptions(request.params.agent)
```
Comment (SawyerHood) [inline packages/agent-runtime/src/acp/bridge/bridge.ts:2009]: 🚨 `slopcop/review` — Medium: run agent discovery in the requested workspace.

The adapter ignores the `model/list` command `cwd`. It builds this agent command from `profile.cwd` only.

OpenCode reads project agents from `.opencode/agents/`. The picker therefore misses agents from the selected environment.

Pass `command.cwd` into the agent command and its cache key. Add a test with two workspace agent sets.

File: apps/app/src/hooks/thread-creation-options/selection-state.ts [hunk 4]
```diff
@@ -281,6 +302,7 @@ export function buildExecutionInputSources({
   if (scope === "component-local") {
     return {
       ...(modelSource ? { model: modelSource } : {}),
+      ...(providerModeSource ? { providerMode: providerModeSource } : {}),
```
Comment (SawyerHood) [inline apps/app/src/hooks/thread-creation-options/selection-state.ts]: 🚨 `slopcop/review` — High: keep creation-only mode data out of existing-thread requests.

The `component-local` scope serves existing threads. Its strict request schema does not accept `providerMode`.

Any changed execution control makes this source object explicit. A send or queue request then returns HTTP 400.

Remove this field from this branch. Add a schema parse test for the complete result.

File: apps/cli/src/commands/provider.ts [hunk 5]
```diff
@@ -106,6 +106,31 @@ export function registerProviderCommands(
...
+    .option("--json", "Print machine-readable JSON output")
+    .action(
+      action(
+        async (
+          providerId: string | undefined,
+          opts: ProviderListCommandOptions,
+        ) => {
+          const serverUrl = getUrl();
+          const sdk = createCliBbSdk(serverUrl);
+          const executionOptions = await sdk.providers.models({
```
Comment (SawyerHood) [inline apps/cli/src/commands/provider.ts]: 🚨 `slopcop/review` — Medium: avoid full model discovery for this mode-only command.

This call gets all models and then discards them. Claude can start an account probe that uses the 30-second host deadline.

ACP providers can also start a temporary child and send one setup request for each model.

Add a mode-only route through the CLI, SDK, server, daemon, and runtime. Skip host work for providers without modes.

File: apps/cli/src/commands/provider.ts [hunk 6]
```diff
@@ -166,3 +191,31 @@ function printModelTable(models: AvailableModel[], providerId?: string): void {
...
+function printModeTable(
+  modes: AvailableProviderMode[],
+  providerId?: string,
+): void {
+  if (providerId) {
+    console.log(`Provider modes for ${providerId}:`);
+  }
+  const rows = modes.map((mode) => [
+    mode.id,
+    mode.displayName,
```
Comment (SawyerHood) [inline apps/cli/src/commands/provider.ts:204]: 🚨 `slopcop/review` — Low: clean untrusted mode labels before terminal output.

An ACP agent or enrolled host controls the mode ID and name. `cli-table3` keeps terminal control sequences and line breaks.

An OSC 52 value can change clipboard data. Other values can show false rows.

Clean all human table cells in `renderBorderlessTable`. Keep JSON output unchanged, and test OSC 52 and newline values.

File: apps/app/src/components/promptbox/ExecutionControls.tsx [hunk 7]
```diff
@@ -58,6 +64,7 @@ export interface ExecutionControlsProps {
   providerRouting?: SystemProvidersQuery;
   provider: ExecutionProviderConfig;
   model: ExecutionModelConfig;
+  providerMode?: ExecutionProviderModeConfig;
```
Comment (SawyerHood) [inline apps/app/src/components/promptbox/ExecutionControls.tsx:67]: 🚨 `slopcop/review` — Medium: update the second production composer for the new agent choice.

`PluginNewThreadComposer` also builds these controls, but it does not pass this optional value.

It does not show the Agent section. Its public `NewThreadRequest` also cannot submit `providerMode`.

Add the field to the plugin contract and request. A shared execution builder can prevent more drift between the composers.

**Review by SawyerHood — COMMENTED** [review]
🚨 **SLOP COP** 🚨 · `review`

ELI5: This change lets bb show OpenCode models and agents. A user can select both before a new thread starts.

I found seven items.

- High: Existing-thread send and queue requests can fail with HTTP 400 after any execution control changes.
- High: A missing ACP mode response can keep a child process after the outer timeout.
- Medium: A mode-only ACP agent loses the synthetic default model and cannot start a thread.
- Medium: Project OpenCode agents are absent because discovery ignores the selected workspace.
- Medium: The plugin new-thread composer does not show or submit the new agent choice.
- Medium: `bb provider modes` runs a complete model probe and can reach the 30-second deadline.
- Low: Terminal control bytes from ACP mode labels reach human CLI output.

The duplicate composer setup caused one of these gaps. A shared execution-control request builder would reduce this drift.

All nine affected Turbo type checks passed. The CLI, bridge, host, configuration, and contract checks passed.

Broad test runs found failures only in unchanged tests. Most failures were load timeouts, and one old stderr test failed twice.

The live picker first reached its deadline while the broad tests ran. A clean API retry returned in 0.08 seconds.

The browser then showed OpenCode models plus `build` and `plan`. It selected Big Pickle and marked `plan` as active.

**Comment by SawyerHood** [comment]
🚨 **SLOP COP** 🚨 · `review`

I am the SlopCop. I am reviewing this pull request for security, code quality, performance, architecture, and end-to-end behavior.

I will use three parallel BB review agents. I will post one final review after I verify their findings.

File: packages/agent-runtime/src/acp/adapter.ts [hunk 8]
```diff
@@ -563,30 +563,39 @@ export function createAcpProviderAdapter(
...
+    envVars?: Record<string, string>;
+  } {
     return {
       command: profile.agentCommand.command,
       args: [...profile.agentCommand.args],
-      ...(profile.cwd !== undefined ? { cwd: profile.cwd } : {}),
+      ...(profile.cwd !== undefined
+        ? { cwd: profile.cwd }
+        : cwd !== undefined
+          ? { cwd }
```
Comment (SawyerHood) [inline packages/agent-runtime/src/acp/adapter.ts:587]: 🚨 `slopcop/review` — High: Repository-scoped discovery can run project code with host secrets.

This line sends the selected repository to the full ACP agent. The bridge also gives the child the daemon environment. OpenCode loads project plug-ins at startup, including files under `.opencode/plugins/`. Therefore, a picker or CLI query can run repository code before a thread starts. The normal ACP permission flow does not control this startup code.

Run discovery in a neutral directory with a limited environment. If project-specific discovery is necessary, require repository trust and disable project plug-ins. [OpenCode plug-in behavior](https://opencode.ai/docs/plugins/)

File: apps/app/src/components/plugin/PluginNewThreadComposer.tsx [hunk 9]
```diff
@@ -676,6 +695,7 @@ export function PluginNewThreadComposer({
       executionInputSources: {
         ...executionInputSources,
         ...seededExecutionInputSources,
+        ...(selectedProviderMode ? { providerMode: "explicit" as const } : {}),
```
Comment (SawyerHood) [inline apps/app/src/components/plugin/PluginNewThreadComposer.tsx:698]: 🚨 `slopcop/review` — High: Keep the provider source with the explicit provider mode.

An unseeded composer can change a Codex project default to OpenCode. The component-local source map omits `providerId`, but this line marks `providerMode` as explicit. The server then discards the requested provider, restores Codex, and rejects the ACP mode.

Mark the changed provider as explicit with the mode. Add an unseeded provider-change test through the server boundary.

File: packages/agent-runtime/src/acp/bridge/bridge.ts [hunk 10]
```diff
@@ -833,6 +854,101 @@ async function loadSessionDiscoveredModels(
...
+      SESSION_MODEL_DISCOVERY_TTL_MS
+  ) {
+    return cachedSessionDiscoveredModes.modes;
+  }
+
+  const childEnv = {
+    ...withoutBridgeRuntimeEnv(process.env),
+    ...(agent.envVars ?? {}),
+  };
+  const connection = createAcpAgentConnection({
```
Comment (SawyerHood) [inline packages/agent-runtime/src/acp/bridge/bridge.ts:880]: 🚨 `slopcop/review` — Medium: Deduplicate active discovery requests.

The cache receives a value only after discovery finishes. Concurrent requests with the same key can all reach this line and start separate ACP processes. Each process can remain active for 30 seconds.

Store an active promise per key. Use one bounded cache for model and mode discovery. Add a short failure cache and a concurrent-request test.

File: apps/app/src/components/pickers/ModelReasoningPicker.tsx [hunk 11]
```diff
@@ -403,6 +419,11 @@ export function ModelReasoningPicker({
     if (!models || models.length === 0) return undefined;
     return models.find((model) => model.isDefault) ?? models[0];
   }, [isPreviewing, previewQuery.data?.models]);
+  const activeProviderModeValue = isPreviewing
```
Comment (SawyerHood) [inline apps/app/src/components/pickers/ModelReasoningPicker.tsx:422]: 🚨 `slopcop/review` — Medium: The preview can show one agent and commit another agent.

The preview always marks the new provider's default mode. Model selection commits only the provider and model. A stored mode can then become active when its identifier is valid for the new provider.

Show the valid stored mode during preview. Alternatively, commit the displayed default mode with the provider and model.

**Review by SawyerHood — COMMENTED** [review]
🚨 **SLOP COP** 🚨 · `review`

ELI5: BB now lets you choose an OpenCode model and agent. It is like choosing a car and its driver. The main path works.

I found four issues that should be fixed before merge.

- High: Repository-scoped discovery can run project plug-ins before a thread starts. The process also receives the daemon environment. See `packages/agent-runtime/src/acp/adapter.ts:587`.
- High: The plugin composer can submit an explicit ACP mode without an explicit provider. The server can restore Codex and reject the mode. See `apps/app/src/components/plugin/PluginNewThreadComposer.tsx:698`.
- Medium: Concurrent mode queries can start duplicate ACP processes. Each process can remain active for 30 seconds. See `packages/agent-runtime/src/acp/bridge/bridge.ts:880`.
- Medium: A provider preview can mark one agent but commit another stored agent. See `apps/app/src/components/pickers/ModelReasoningPicker.tsx:422`.

The discovery code has two similar process-start paths. One shared path should own trust, environment, timeouts, cleanup, and active-request caching.

The plugin composer also duplicates root composer setup. A shared new-thread request builder would prevent source-map differences.

Validation:

- I reviewed the exact head SHA.
- The browser selected Big Pickle, agent `plan`, and `accept-edits`. OpenCode returned `ok`.
- The stored event included `providerMode: plan`.
- All nine affected Turbo type checks passed.
- Focused app, server, host, CLI, and ACP tests passed.
- One unchanged stderr-tail test failed locally twice. GitHub package tests passed.
- All current GitHub checks passed.
- The daemon protocol version increased to 96.

I posted detailed inline comments for each issue. I used a comment review only.

**Comment by iamhenry** [comment]
awww i was really looking forward to this

**Comment by SawyerHood** [comment]
@iamhenry I have another PR coming. This one was just slop lol

**Comment by iamhenry** [comment]
haha thx! appreciate that!
